Documentation/using-corert/reflection-in-aot-mode.md

```diff
@@ -0,0 +1,65 @@
+
+As a motivating example, consider this program:
+
+```csharp
+class Program
+{
+    public static void Main()
+    {
+        Console.Write("Name of type: ");
+        string typeName = Console.ReadLine();
+        if (String.IsNullOrEmpty(typeName))
+            return;
+
+        Console.Write("Name of method: ");
+        string methodName = Console.ReadLine();
+
+        Type.GetType(typeName).GetMethod(methodName).Invoke(null, null);
+    }
+
+    public static void SayHello()
+    {
+        Console.WriteLine("Hello!");
+    }
+}
+
+```
+
+The above program lets the user invoke any parameterless public static method on any type. For the naive compilation algorithm above, this program would work great for input strings `Program` and `Main` because the algorithm included method `Main` in the final executable. The program wouldn't work so great for inputs `Program` and `SayHello`*, because method `SayHello` wasn't called from anywhere. For the naive algorithm, the only way to fix the program for inputs `Program` and `SayHello` is to add a call to `SayHello` in `Main`.
+
+> * The behavior for `Type.GetMethod` on type `Program` and method `SayHello` would be to return `null` if `SayHello` wasn't compiled. The reason for this is that `Type.GetMethod` is documented to return `null` if there's no method with a given name, and for the purposes of the program, `SayHello` doesn't exist. The compiler could remember there used to be such method and write the information in the executable, but that would raise additional questions about whether the uncompiled method should be included in a `Type.GetMethods` call.
+
+While the example program above is not practical in reality, similar patterns exist in e.g. reflection based serialization and deserialization libraries that access members based on their names that could be e.g. downloaded from the internet.
+
+The dynamic nature of reflection doesn't pose a problem for .NET platforms that have a JIT compiler or an interpreter (and don't attemt to remove "unused" parts of the program, such as when using the IL Linker). JIT and interpreter in CoreRT are optional components in early experimental stages and cannot be used to solve the problem at this time. There are also scenarios that can't afford the overhead of a JIT or an interpreter, so a solution that can support reflection in a fully ahead of time environment is still desirable. An environment without a JIT or an interpreter is closer to the environment Go or Rust developers are used to (and environment that is more lightweight, with predictable code generation characteristics).
+
+## Solving reflection in full AOT mode ##
+
+The solution to reflection is about establishing what parts of the program can be reached dynamically and making sure their metadata and code is available at runtime.
```

Comment:
It might be useful to list which option (if there is one) relates to each of the "Assume .." mode, i.e. `RootAllApplicationAssemblies`, `IlcGenerateCompleteTypeMetadata`, etc and indicate if that mode is the default or not?